### PR TITLE
Add CookiePolicy and consent UI to the template

### PR DESCRIPTION
**Pull request by Tratcher**
@DamianEdwards This is a preview of adding CookiePolicy to the templates. We should not merge it until the corresponding consent UI is available. @glennc @blowdart who's working on the UI?

We also need to add this to most of the other templates.

### CODE REVIEW

**Comment by javiercn** [comment]
You need to update the razorpages one too (and file a bug for the F# templates)

**Comment by Tratcher** [comment]
I know, I wanted Damian to see the Startup code first.

**Comment by DamianEdwards** [comment]
Please see candidate UI for inclusion in this PR at https://github.com/DamianEdwards/CookieConsentUI

**Comment by Tratcher** [comment]
#208

**Comment by Tratcher** [comment]
I've incorporated the UI provided by Damian into the Razor and MVC templates.

Waiting on https://github.com/aspnet/Security/pull/1595 to clean up the cookie generation.

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+﻿@using Microsoft.Extensions.Options
+@using Microsoft.AspNetCore.Builder
+@using Microsoft.AspNetCore.Http.Features
+@using Microsoft.Net.Http.Headers
+
+@inject IOptions<CookiePolicyOptions> CookiePolicy
+
+@{
+    var showBanner = (!Context.Features.Get<ITrackingConsentFeature>()?.CanTrack ?? false) && CookiePolicy.Value.ConsentCookie != null;
```
Comment (javiercn) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml]: What's the actual logic here? what does CookiePolicyOptions.Value.ConsentCookie != null mean?

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    // Might be nice if the feature/builder could simply generate and return the cookie string for use in JS for me
+
+    var cookieName = CookiePolicy.Value.ConsentCookie?.Name;
+    var cookieValue = "yes"; // NOTE: Replace this with value from incoming public constant, see https://github.com/aspnet/Security/issues/1590
```
Comment (javiercn) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml]: Is this a temporary thing? (the bug is marked as done)

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        Path = options.Path,
+        Expires = options.Expires,
+        MaxAge = options.MaxAge,
+        Secure = options.Secure,
+        SameSite = (Microsoft.Net.Http.Headers.SameSiteMode)options.SameSite,
+        HttpOnly = options.HttpOnly
+    };
+
+    var cookieString = setCookieHeaderValue.ToString();
+}
```
Comment (javiercn) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml]: All the stuff above should not be in the view. I would suggest all this logic is encapsulated in a service somewhere.

Comment (Tratcher) [reply src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Shared/_CookieConsentPartial.cshtml]: All of the things you commented on are going away as soon as I can consume the new feature from the linked issue. Universe is still building.

**Comment by Tratcher** [comment]
Ready for final review.

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs [hunk 4]
```diff
@@ -49,6 +49,11 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            services.Configure<CookiePolicyOptions>(options =>
+            {
+                options.CheckConsentNeeded = context => true;
```
Comment (DamianEdwards) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs]: Can we add a code comment here please:
``` c#
// This lambda determines whether user consent for non-essential cookies is needed for a given request.
options.CheckConsentNeeded = context => true;
```

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Startup.cs [hunk 5]
```diff
@@ -72,6 +72,11 @@ public void ConfigureServices(IServiceCollection services)
             .AddCookie();
 
 #endif
+            services.Configure<CookiePolicyOptions>(options =>
+            {
+                options.CheckConsentNeeded = context => true;
```
Comment (DamianEdwards) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/StarterWeb-CSharp/Startup.cs]: Comment from above.

**Pull request merged**